rascaline/src/calculators/atomic_composition.rs

```diff
@@ -0,0 +1,269 @@
+            samples.push(Arc::new(builder.finish()));
+        }
+
+        return Ok(samples);
+    }
+
+    fn supports_gradient(&self, parameter: &str) -> bool {
+        match parameter {
+            "positions" => true,
+            "cell" => true,
+            _ => false,
+        }
+    }
+
+    fn positions_gradient_samples(
+        &self,
+        keys: &Labels,
+        samples: &[Arc<Labels>],
+        systems: &mut [Box<dyn System>],
+    ) -> Result<Vec<Arc<Labels>>, Error> {
+        debug_assert_eq!(keys.count(), samples.len());
+        let mut gradient_samples = Vec::new();
+        for ([species_center_key], samples) in keys.iter_fixed_size().zip(samples) {
+            let mut builder = LabelsBuilder::new(vec!["sample", "structure", "atom"]);
+            if self.per_structure {
+                for (system_i, system) in systems.iter_mut().enumerate() {
+                    let species = system.species()?;
+
+                    for (center_i, &species_center_sys) in species.iter().enumerate() {
+                        if species_center_key.i32() == species_center_sys {
+                            builder.add(&[system_i, system_i, center_i]);
+                        }
+                    }
+                }
+            } else {
+                for (sample_i, [structure_i, center_i]) in samples.iter_fixed_size().enumerate() {
+                    builder.add(&[sample_i, structure_i.usize(), center_i.usize()]);
+                }
+            }
+            gradient_samples.push(Arc::new(builder.finish()));
+        }
+        return Ok(gradient_samples);
+    }
+
+    fn components(&self, keys: &Labels) -> Vec<Vec<Arc<Labels>>> {
+        return vec![Vec::new(); keys.count()];
+    }
+
+    fn properties_names(&self) -> Vec<&str> {
+        return vec!["count"];
+    }
+
+    fn properties(&self, keys: &Labels) -> Vec<Arc<Labels>> {
+        let mut properties = LabelsBuilder::new(self.properties_names());
+        properties.add(&[0]);
+        let properties = Arc::new(properties.finish());
+
+        return vec![properties; keys.count()];
+    }
+
+    fn compute(
+        &mut self,
+        systems: &mut [Box<dyn System>],
+        descriptor: &mut TensorMap,
+    ) -> Result<(), Error> {
+        assert_eq!(descriptor.keys().names(), ["species_center"]);
+
+        for (key, mut block) in descriptor.iter_mut() {
+            let values = block.values_mut();
+            let array = values.data.as_array_mut();
+
+            if self.per_structure {
+                let species = key[0].i32();
+                for (system_i, system) in systems.iter_mut().enumerate() {
+                    let all_species = system.species()?;
+                    let mut n_species = 0.;
+
+                    for &s in all_species {
+                        if s == species {
+                            n_species += 1.;
+                        }
+                    }
+                    array[[system_i, 0]] = n_species;
```

Comment:
This will fail with samples/properties selection. The user could remove the "count" property (would be stupid, but meh), or have samples in a different order than the systems.

Reply:
So we do a loop over properties like

```rust
for (property_i, property) in values.properties.iter().enumerate() {
}
```

and only fill if  `property_i  == 0`?

Reply:
Okay, I changed the code. Looks still quite clean and simple.